stats/opentelemetry/internal/tracing/custom_carrier.go

```diff
@@ -0,0 +1,103 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package tracing implements the OpenTelemetry carrier for context propagation
+// in gRPC tracing.
+package tracing
+
+import (
+	"context"
+	"errors"
+
+	"go.opentelemetry.io/otel/propagation"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+)
+
+// GRPCTraceBinHeaderKey is the gRPC metadata header key `grpc-trace-bin` used
+// to propagate trace context in binary format.
+const GRPCTraceBinHeaderKey = "grpc-trace-bin"
+
+// CustomCarrier is a TextMapCarrier that uses gRPC context to store and
+// retrieve any propagated key-value pairs in text format along with binary
+// format for `grpc-trace-bin` header
+type CustomCarrier struct {
+	propagation.TextMapCarrier
+
+	Ctx *context.Context
+}
+
+// NewCustomCarrier creates a new CustomMapCarrier with
+// the given context.
+func NewCustomCarrier(ctx context.Context) CustomCarrier {
+	return CustomCarrier{
+		Ctx: &ctx,
+	}
+}
+
+// Get returns the string value associated with the passed key from the gRPC
+// context.
+func (c CustomCarrier) Get(key string) string {
+	md, ok := metadata.FromIncomingContext(*c.Ctx)
+	if !ok {
+		return ""
+	}
+	values := md.Get(key)
+	if len(values) == 0 {
+		return ""
+	}
+	return values[0]
+}
+
+// Set stores the key-value pair in string format in the gRPC context.
+// If the key already exists, its value will be overwritten.
+func (c CustomCarrier) Set(key, value string) {
+	md, ok := metadata.FromOutgoingContext(*c.Ctx)
+	if !ok {
+		md = metadata.MD{}
+	}
+	md.Set(key, value)
+	*c.Ctx = metadata.NewOutgoingContext(*c.Ctx, md)
+}
+
+// GetBinary returns the binary value from the gRPC context in the incoming RPC,
+// associated with the header `grpc-trace-bin`.
+func (c CustomCarrier) GetBinary() ([]byte, error) {
+	values := stats.Trace(*c.Ctx)
+	if len(values) == 0 {
+		return nil, errors.New("`grpc-trace-bin` header not found")
```

Comment:
Is this the correct error message? I think both in the case it's not present in md and if it's set to an empty byte string it'll be not found. Is an empty byte string distinct from nil in the return? Is the empty byte string a valid thing to transmit for this header? Do we want this to return an error at all?

Reply:
Removed returning error from here to be consistent with `Get(..)`

Reply:
I'm fine with the error. What is the intended usage of this function? Do you intended to treat not set and set to an empty string equivalent? Is empty string a valid key for this? In that case you should make the error message fail with that. But that relates to the intended usage of this, and the failure modes (i.e. what happens in the error case).

Reply:
Here we are only looking trace context which needs to conform to following struct. So empty byte is essentially nil.

```
// Fields:
        //
        // TraceId: (field_id = 0, len = 16, default = "0000000000000000") -    
        // 16-byte array representing the trace_id.
        //
        // SpanId: (field_id = 1, len = 8, default = "00000000") - 8-byte array   
        // representing the span_id.
        //
        // TraceOptions: (field_id = 2, len = 1, default = "0") - 1-byte array 
           representing the trace_options.
```

Reply:
> Is empty string a valid key for this?

Binary methods will not deal with string keys because we are using existing stats package https://github.com/grpc/grpc-go/blob/d9d8f342b7ee5b405e3c97fb51cda75643ce7cf1/stats/stats.go#L316 to set bytes which sets a custom struct key